rfc(decision): Replay Issue Creation (#92)
getsentry/rfcs: text/0092-replay-issue-creation.md

- Start Date: 2023-05-15
- RFC Type: decision
- RFC PR: https://github.com/getsentry/rfcs/pull/92
- RFC Status: draft

# Summary

We want to detect certain categories of issues only available through the Session Replay product. The first of these issues is the "Slow Click" detection issue.

# Motivation

1. Inform the long-term strategy for detecting and ingesting Session Replay generated issues.
2. Provide actionable feedback to developers that could not otherwise be captured in a performance span or error event.
3. Increase product awareness in free-tier customers and encourage adoption.

> # Options Considered

Reviewer: I'm trying to think of a way to make sure we're looking at an exhaustive list of options, even if some are bad. 

here's where i'm starting from:

| SDK | Network data | Relay & Ingest | Result |
| --- | --- | --- | --- |
| @sentry/javascript | error payload (call `captureException()` in browser)| normal processing | error quota risk |
| @sentry/javascript | click detection (same payload as in experiment) | server-side detector needed | - | 
| @sentry/replay | existing breadcrumbs + rrweb events | normal relay processing | server-side detector needed | replay must be sampled |
| @sentry/replay | click detection breadcrumb (same payload as in experiment) | server-side detector needed | replay must be sampled |

Reviewer 2: @ryan953 @cmanallen Could we add a column here that is "Product Experience" or "Product Implications" so we can visualize what this means to our users?

> We want to detect certain categories of issues only available through the Session Replay product. These issues can only be detected on the SDK. The Replay back-end will never have enough data to find these issues. For that reason this is primarily an SDK driven workload. The question is: what role should the Replay back-end have in Replay issue creation? Should the Replay SDK use the Replay back-end to generate new issues or should the SDK generate those issues through a generic, non-replay-specific interface?

Reviewer 2: > We want to detect certain categories of issues only available through the Session Replay product.

Since we _can_ detected these categories of issues without the help of the Replay SDK all together, we should probably reword the "issues only available through the Session Replay product." language. How about, "issues that can greatly benefit from videos and context in the Session Replay product." or similar?

> We want to detect certain categories of issues only available through the Session Replay product. These issues can only be detected on the SDK. The Replay back-end will never have enough data to find these issues. For that reason this is primarily an SDK driven workload. The question is: what role should the Replay back-end have in Replay issue creation? Should the Replay SDK use the Replay back-end to generate new issues or should the SDK generate those issues through a generic, non-replay-specific interface?

Reviewer 2: ```suggestion
We want to detect certain categories of issues only available through the Session Replay product. These issues can only be detected on the SDK. The Replay back-end will never have enough data to find these issues. For that reason this is primarily an SDK driven workload. The question is: what role should the Replay back-end have in Replay issue creation? Should the Replay SDK use the Replay back-end to generate new issues or should the SDK generate those issues through a generic, non-replay-specific interface? Each option would have significantly different product implications that will be discussed below.
```

> We want to detect certain categories of issues only available through the Session Replay product. These issues can only be detected on the SDK. The Replay back-end will never have enough data to find these issues. For that reason this is primarily an SDK driven workload. The question is: what role should the Replay back-end have in Replay issue creation? Should the Replay SDK use the Replay back-end to generate new issues or should the SDK generate those issues through a generic, non-replay-specific interface?

Reviewer 3: When we first worked on performance issues the initial spike was doing them via sdk. But there was a lot of pushback to move things to the backend, since implementing this logic across N sdks isn't really a scalable solution.

Maybe this isn't such a big deal if this only applies to the JS world though? Either way it might be worth checking that upper management isn't going to veto this approach later on.

Author: > Maybe this isn't such a big deal if this only applies to the JS world though? 

Yeah I'm hoping thats the case.

Unfortunately this work has to be done on the SDK due to the large sizes of replays and the large time differences between segments (segments can refer back to any previous segment).  This is probably something that should be documented in the RFC :P

> 3. Increase product awareness in free-tier customers and encourage adoption.

Reviewer 2: ```suggestion
3. Increase product awareness in platform customers and encourage adoption.
```

> ### Option 1: SDK

Reviewer 2: Can we have the option names match the wording you used in the Summary section? Or vice versa. It would make it a little easier to follow~

"Should the Replay SDK use the Replay back-end to generate new issues or should the SDK generate those issues through a generic, non-replay-specific interface?"

When the SDK encounters a "replay issue" it will make an HTTP request to a generic interface which will handle the issue creation process.

> **Pros:**

Reviewer 2: For each option, I would split out the Pros/Cons from engineering POV vs. product POV. (I can fill in Product ✨ ) 

A clear "Pro" to this solution is we'd be exposing Dead Click Issues to much much larger customer segment when compared to only Replay customers, for example. 😃

Author: Added.  I left it as a single pros/cons table for now.  If we start to accumulate a lot of items we can split it.

1. Can be tested in isolation without impacting existing production services.

**Cons:**

> 1. Uses quota.

Reviewer 2: I assume you mean errors quota here. It would be good to state that specifically for the revised RFC content.

As an aside, is it possible to do Option 1 and exclude it from consuming errors quota?

Reviewer 2: ^ If the answer is 'yes' to the above question, we'll have to modify the pros/cons list

2. Unclear if we're able to use dynamic thresholds.
   - E.g. "Experienced 10 occurences in the past hour".

> 3. The replay containing the issue could be sampled.

Reviewer 2: ```suggestion
3. The replay containing the issue could not be sampled.
```

### Option 2: Ingest

The SDK publishes a "replay issue" to the Replay back-end. The back-end will decide to process the issue or not.

**Pros:**

> 1. Does not use quota.

Reviewer 2: ```suggestion
1. Does not use errors quota.
```

**Pros:**

1. Can be tested in isolation without impacting existing production services.

**Cons:**

1. Uses quota.
2. There's overhead associated with submitting a new issue through an HTTP request.
3. Unable to use dynamic thresholds.
   - E.g. "Experienced 10 occurences in the past hour".
4. The replay containing the slow click could be sampled.

### Option 2: Ingest

We create a set of detectors on the backend. When a detector is activated we publish to the issue creation topic.

The SDK will update the recording breadcrumbs to include a "Slow Click" breadcrumb event type. The ingest server will use that event type to make an issue creation decision.

**Pros:**

1. Does not use quota.
2. Overhead is low.
   - Publishing to a Kafka consumer can happen asynchronously.
3. We can use dynamic thresholds.
   - E.g. "Experienced 10 occurences in the past hour".
4. Replay is guaranteed to be sampled.

**Cons:**

1. Poor rollout could impact service availability during testing period.

> 2. Requires coordination between the SDK and Ingest to create new issue types.

Author: Profiling does it in the back end.  Performance I'm not sure exactly.  I know they have a lot of back end detectors so at least some portion is handled there.
